src/test/fio/fio_ceph_objectstore.cc

```diff
@@ -0,0 +1,393 @@
+        events++;
+      }
+    }
+    if (events >= min)
+      break;
+    usleep(100);
+  } while (1);
+
+  return events;
+}
+
+/// completion context for ObjectStore::queue_transaction()
+class UnitComplete : public Context {
+  io_u* u;
+ public:
+  UnitComplete(io_u* u) : u(u) {}
+  void finish(int r) {
+    // mark the pointer to indicate completion for fio_ceph_os_getevents()
+    u->engine_data = reinterpret_cast<void*>(1ull);
+  }
+};
+
+int fio_ceph_os_queue(thread_data* td, io_u* u)
+{
+  fio_ro_check(td, u);
+
+  auto job = static_cast<Job*>(td->io_ops_data);
+  auto& object = job->objects[u->file->engine_data];
+  auto& coll = object.coll;
+  auto& os = job->engine->os;
+
+  if (u->ddir == DDIR_WRITE) {
+    bufferlist bl;
+    bl.push_back(buffer::create_static(u->xfer_buflen,
+                                       static_cast<char*>(u->xfer_buf)));
+    // enqueue a write transaction on the collection's sequencer
+    ObjectStore::Transaction t;
+    t.write(coll.cid, object.oid, u->offset, u->xfer_buflen, bl);
+    os->queue_transaction(&coll.sequencer, std::move(t), new UnitComplete(u));
+    return FIO_Q_QUEUED;
+  }
+
+  if (u->ddir == DDIR_READ) {
+    // ObjectStore reads are synchronous, so make the call and return COMPLETED
+    bufferlist bl;
+    int r = os->read(coll.cid, object.oid, u->offset, u->xfer_buflen, bl);
+    if (r < 0) {
+      u->error = r;
+      td_verror(td, u->error, "xfer");
+    } else {
+      bl.copy(0, bl.length(), static_cast<char*>(u->xfer_buf));
+      u->resid = u->xfer_buflen - r;
+    }
+    return FIO_Q_COMPLETED;
+  }
+
+  derr << "WARNING: Only DDIR_READ and DDIR_WRITE are supported!" << dendl;
+  u->error = -EINVAL;
+  td_verror(td, u->error, "xfer");
+  return FIO_Q_COMPLETED;
+}
+
+int fio_ceph_os_commit(thread_data* td)
+{
+  // commit() allows the engine to batch up queued requests to be submitted all
+  // at once. it would be natural for queue() to collect transactions in a list,
+  // and use commit() to pass them all to ObjectStore::queue_transactions(). but
+  // because we spread objects over multiple collections, we a) need to use a
+  // different sequencer for each collection, and b) are less likely to see a
+  // benefit from batching requests within a collection
+  return 0;
+}
+
+// open/close are noops. we set the FIO_DISKLESSIO flag in ioengine_ops to
+// prevent fio from creating the files
+int fio_ceph_os_open(thread_data* td, fio_file* f) { return 0; }
+int fio_ceph_os_close(thread_data* td, fio_file* f) { return 0; }
+
+int fio_ceph_os_io_u_init(thread_data* td, io_u* u)
+{
+  // no data is allocated, we just use the pointer as a boolean 'completed' flag
+  u->engine_data = nullptr;
+  return 0;
+}
+
+void fio_ceph_os_io_u_free(thread_data* td, io_u* u)
+{
+  u->engine_data = nullptr;
+}
+
+
+// ioengine_ops for get_ioengine()
+struct ceph_ioengine : public ioengine_ops {
+  ceph_ioengine() : ioengine_ops({}) {
+    strncpy(name, "ceph-os", sizeof(name));
```

Comment:
Recent commit to fio causes build failure here. Now they have "const char* name;" in ioengine_ops and that's inappropriate for strncpy.